app/controllers/katello/api/v2/upstream_subscriptions_controller.rb

```diff
@@ -0,0 +1,22 @@
+module Katello
+  class Api::V2::UpstreamSubscriptionsController < Api::V2::ApiController
+    before_action :find_organization
+
+    resource_description do
+      description "Subscriptions management on the CDN."
+      api_version 'v2'
+    end
+
+    api :GET, "/organizations/:organization_id/upstream_subscriptions", N_("List available subscriptions from the CDN")
+    param :organization_id, :number, :desc => N_("Organization ID"), :required => true
```

Comment:
This works, but I think we need some more functionality. Let me know if you envisioned these things being added in a later commit btw:

The way this will be used by the UI is always (afaik) going to be in the context of the upstream consumer from which the manifest was initially exported so I think we need to accept the consumer UUID parameter here to filter by, and maybe it should even be required.  Consumer#available_subscriptions has an example. We can forward those same params to hosted RHSM.

From a performance perspective I don't think retrieving the entire dataset is going to work. We have a few options: Candlepin supports 'include' and 'exclude' params which we can use to whittle down the attributes returned in the response. We will only need a few of the attributes from the pool to render our view. There is also paging: http://www.candlepinproject.org/docs/candlepin/pagination.html page & per_page seem most relevant.

What do you think?

Reply:
Yeah, the design doc is pretty unclear about the output columns/pagination/etc. and which params should be able to be passed. I left it untouched because of this, but I see no reason why we can't make these decisions on this PR. I like all your suggestions! /me gets to work

Reply:
@jturel Could you ellaborate on the required UUID param? I don't really understand why this would be required when our API users have no concept of candlepin consumers afaik. The organization_id is the only required param here because it (potentially) contains the manifest. I am unable to find the example you mentioned in Consumer#available_subscriptions - perhaps you could share a link?

Reply:
Yes! My thinking is: the imported manifest comes from an upstream RHSM consumer. The new UI we are building is manipulating that very consumer - changing subscription quantities, adding, and removing them altogether. While the user is not thinking in those terms it means that all of those API calls should be within the context of that consumer. We should be able to get its UUID from the identity cert if we don't have it stored elsewhere. The user shouldn't have to provide it themselves which may have made my comment confusing...

I was referring to this code: https://github.com/Katello/katello/pull/7197/files#diff-4e53ba860fcaf2eb36f0da5a96751337R83

The reason that we need to send the UUID is because we don't actually want to show all of the pools on the account - some of them may not have any entitlements available. Passing in the UUID will give us what can actually be attached to the consumer and make it into the manifest. IOW I'm not sure we have a use case to list all of the pools on the account regardless of whether they are fully consumed or not.

Sorry that was so long - hope it helps.

Reply:
I think I'm doing this already: https://github.com/Katello/katello/pull/7197/commits/09e012f7a63b186a48bd0d570966dedf79ffe99f#diff-d4ada19367f2e37f2adeafa9de778bccR428

Reply:
Ah, I do see that. It looks like the controller is using the UpstreamPool resource rather than UpstreamConsumer and so we're getting all the pools.

re: account - yep :+1: 

@jlsherrill can you validate my previous comment before Andrew makes the change?

Reply:
Right, we're getting all the pools for that owner. IOW, show me all the pools I could attach to my manifest. I think that's what we want?

Reply:
Not quite. All the pools for the owner isn't the same as what could be attached to the consumer. Maybe some of those pools have 0 entitlements available => they can not be attached.

Giving it the consumer UUID will give us a valid list of pools that can be attached which is a subset of all pools on the account/owner.

Reply:
If i'm reading correctly, i think that jturel is suggesting we do something like this:
https://github.com/Katello/katello/blob/master/app/lib/katello/resources/candlepin.rb#L200-L204

where we are limiting to subs available to the consumer.  I'm not 100% sure if we should be using this or not, as there was some discussion around whether we should allow you to add multiple instances of the same pool (which i think the portal allows you to do in the new ui?)  Maybe we need to compare the results of the two.

Reply:
@jturel Yeah, and i thought i asked about the zero case specifically and was told that those should still show up (my memory could be faulty though).  Maybe we should add a task to examine the pros/cons of both of these methods and leave it as it for now?

Reply:
partially because i think it will be more obvious once we can actually use the tool

Reply:
The portal would have done that before new UI too. If we want to prevent double attach we'll have to do it on our side no matter how we invoke the API as far as I can tell. Comparing what is already attached to the results from upstream...

Reply:
:+1: to leaving for now and adjusting as we work out the feature.

Reply:
@akofink i do think we need to make sure to revisit though, so if you could file a card or issue, that'd be awesome!

Reply:
http://projects.theforeman.org/issues/22710